src/www/Inset.ts

```diff
@@ -0,0 +1,266 @@
+   limitations under the License.
+*/
+
+export interface IInsetAPI {
+    addListener: (callback: IInsetCallbackFunc) => void;
+    removeListener: (callback: IInsetCallbackFunc) => void;
+    getInsets: () => IInset;
+}
+
+declare global {
+    interface ITotalpave {
+        Insets: IInsetAPI;
+    }
+    interface Window {
+        totalpave: ITotalpave;
+    }
+}
+
+export const SERVICE_NAME: string = "Insets";
+
+export interface IInset {
+    top: number;
+    right: number;
+    bottom: number;
+    left: number;
+}
+
+interface IInsetEvent<T = unknown> {
+    type: 'init' | 'update';
+    data: T;
+}
+
+type IInsetInitEvent = IInsetEvent<string>;
+interface IInsetUpdateEvent extends IInsetEvent<IInset> {
+    id: string;
+}
+
+export type IInsetCallbackFunc = (inset: IInset) => void;
+
+export interface IInsetConfiguration {
+    /**
+     * A bit mask of InsetMask
+     * 
+     * @defaults DISPLAY_CUTOUT | SYSTEM_BARS
+     */
+    mask?: number;
+
+    /**
+     * If true, includes rounded corners in the inset information
+     * Only available on Android API 31 ("S") and later.
+     * 
+     * @defaults true
+     */
+    includeRoundedCorners?: boolean;
+}
+
+export class Inset { 
+    private static $instances: Map<string, Inset> = new Map();
+
+    private $currentInsets: IInset;
+    private $listeners: IInsetCallbackFunc[];
+    private $id: string;
+
+    private constructor(id: string) {
+        this.$id = id;
+        this.$listeners = [];
+        this.$currentInsets = {
+            top: 0,
+            left: 0,
+            right: 0,
+            bottom: 0
+        };
+    }
+
+    /**
+     * Gets the native identifier
+     * 
+     * @returns 
+     */
+    public getID(): string {
+        return this.$id;
+    }
+
+    /**
+     * Gets the last emitted inset information
+     * 
+     * @returns
+     */
+    public getInsets(): IInset {
+        return this.$currentInsets;
+    }
+
+    /**
+     * See the static Inset.free method for details
+     * 
+     * This is the equivilant of calling Inset.free(insetInstance)
+     * 
+     * @returns 
+     */
+    public async free(): Promise<void> {
+        return await Inset.free(this);
+    }
+
+    /**
+     * Adds a listener to this inset configuration.
+     * 
+     * Note that this may fire even if nothing has actually
+     * changed.
+     * 
+     * Retain the listener reference to remove it later if
+     * necessary.
+     * 
+     * @param listener 
+     */
+    public addListener(listener: IInsetCallbackFunc): void {
+        this.$listeners.push(listener);
+        listener(this.$currentInsets);
+    }
+
+    /**
+     * Frees the listener reference
+     * 
+     * @param listener 
+     */
+    public removeListener(listener: IInsetCallbackFunc): void {
+        let idx: number = this.$listeners.indexOf(listener);
+        if (idx > -1) {
+            this.$listeners.splice(idx, 1);
+        }
+    }
+
+    public onUpdate(insets: IInset): void {
```

Comment:
Why is this public?

Reply:
I think TS will allow it to be private since it's coming from the same file even if it's a static method but it's more or less schematics.

I think I'll change it to a private method however.